### ISSUE TITLE
 the difference between soft-topk and hard-topk

### ISSUE THREAD

**Issue by lijun2005** [HySonLab/SAMSA#1]
An excellent piece of work, with well-founded theory and highly standardized code!!!

 I have a few minor questions I'd like to ask the author.
1. Could you elaborate more on the difference between soft-topk and hard-topk selection? In the code, it seems that the difference is only in the forward propagation, while the backward part of the code is almost identical.

2. What is the role of the mask when obtaining the importance score?"
   ```python
    def forward(self,
                x: torch.Tensor,
                mask: torch.Tensor):
        b, n = x.shape[0], x.shape[1]
        ret = self.transf(x + self.norm(self.importance_score(x))) + ((1 - mask) * (-1e9)).reshape(b, n, 1)
        return ret
     ```

**Comment by minh1409** [comment]
Sure, please clarify if it is not clear somewhere.
1.  Our idea to optimize the selection algorithm is to give it a custom gradient function. Otherwise, the gradient of selection algorithm is zero. Soft-top-k is designed as the proxy to optimize for hard-top-k; therefore, by definition, the soft top-k and hard top-k have the same gradient functions and different forward functions.
2. We don't want to choose from masked elements, for example, different sequence lengths. So by settings the importance scores of those do not exists:
  - The forward pass of top-k using argsort will put that in the last positions
  - The backward pass of top-k using sigmoid will put zeros
 This is combined with a learnable padding of size (k or 2k) to ensure the masked elements are not taken in consideration.

**Comment by lijun2005** [comment]
Thank you very much for your prompt response. I have understood the two issues mentioned above. I have a small additional question: What is the typical initial value for the temperature in the sampling module? 
I notice that the temperature in your config file seems to be set to 0.1.  Is this value determined through ablation experiments?

```python
class DSZRC(nn.Module):
    def __init__(self,
                 nhead: int, 
                 n_sampled_token: int,
                 n_dimension_x: int,
                 n_dimension_qk: int,
                 temperature,
                 hard):
        super().__init__()
        self.nhead = nhead
        self.n_sampled_token = n_sampled_token
        # Define a linear transformation to compute the importance score for each token
        # This layer maps from n_dimension to nhead
        self.importance_score = ImportantScore(nhead, n_dimension_x)
        self.d_model_head = n_dimension_x // nhead
        self.extra_token_score_x = nn.Parameter(torch.randn(1, 2 * n_sampled_token, nhead), requires_grad=True)
        self.extra_token_q = nn.Parameter(torch.randn(1, nhead, 2 * n_sampled_token, n_dimension_qk // nhead), requires_grad=True)
        self.temperature = nn.Parameter(torch.ones(1, nhead, 1, 1) * temperature, requires_grad=True)
        self.grad_approx = nn.Parameter(torch.zeros(1, nhead, 1, n_dimension_qk // nhead), requires_grad=True)
        self.hard = hard

```

**Comment by minh1409** [comment]
I only experimented with 0.1 (soft) and 1.0 (hard) so I don't know;
In the implementation, large temperatures have their gradient clipped so any temperature is probably "trainable":
```python
      sigmoid_grad = p * (1 - p) / temperature  # derivative of sigmoid
      max_sigmoid_grad = torch.amax(sigmoid_grad, dim=(0, 2, 3), keepdim=True)[0]
      sigmoid_grad = sigmoid_grad * torch.clamp(max_sigmoid_grad, 0.0, 0.25) / max_sigmoid_grad
```
Without this, 0.1 is somewhat hard to learn, empirically.

**Comment by lijun2005** [comment]
OK， Thanks



---Original---
From: ***@***.***&gt;
Date: Sun, Sep 1, 2024 22:20 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [HySonLab/SAMSA]  the difference between soft-topk and hard-topk(Issue #1)




 
I only experimented with 0.1 (soft) and 1.0 (hard) so I don't know;
 In the implementation, large temperatures have their gradient clipped so any temperature is probably "trainable":
       sigmoid_grad = p * (1 - p) / temperature  # derivative of sigmoid       max_sigmoid_grad = torch.amax(sigmoid_grad, dim=(0, 2, 3), keepdim=True)[0]       sigmoid_grad = sigmoid_grad * torch.clamp(max_sigmoid_grad, 0.0, 0.25) / max_sigmoid_grad
 
Without this, 0.1 is somewhat hard to learn, empirically.

**Comment by lijun2005** [comment]
I am noticing an inconsistency between the train and eval phases in the code.
In the original code:
```python
       importance_scores = self.importance_score(x, mask)
        importance_scores = torch.cat([importance_scores, self.extra_token_score_x.expand(x.shape[0], -1, -1)], dim=1)
        importance_scores = get_gumbel(importance_scores)

        if self.training is True or self.hard is False:
            # Argsort and gather
            importance_scores, indices = torch.topk(importance_scores, self.n_sampled_token * 2, dim=1, largest=True, sorted=True)
            importance_scores = importance_scores.transpose(1, 2).unsqueeze(-1)
            indices = indices.transpose(1, 2).unsqueeze(-1)
            q = torch.take_along_dim(q, indices=indices, dim=2)
            q_top, q_bot = torch.split(q, split_size_or_sections=[self.n_sampled_token, self.n_sampled_token], dim=2)
            imp_top, imp_bot = torch.split(
                importance_scores, 
                split_size_or_sections=[self.n_sampled_token, importance_scores.shape[2] - self.n_sampled_token],
                dim=2)
            random_indices = torch.randn(x.shape[0], self.nhead, self.n_sampled_token, 1, device=x.device).argsort(dim=2)
            q_bot = torch.take_along_dim(q_bot, indices=random_indices, dim=2)
            imp_bot = torch.take_along_dim(imp_bot, indices=random_indices, dim=2)
            if self.hard is True:
                return HardChooseDenseGrad.apply(imp_top, imp_bot, q_top, q_bot, self.temperature, self.grad_approx)
            else:
                return SoftChooseDenseGrad.apply(imp_top, imp_bot, q_top, q_bot, self.temperature)
        
        else:
            _, indices_top = torch.topk(importance_scores, self.n_sampled_token, dim=1, largest=True, sorted=False)
            indices_top = indices_top.transpose(1, 2).unsqueeze(-1)
            q_top = torch.take_along_dim(q, indices=indices_top, dim=2)
            return q_top
```
There are two issues：
1. ``` importance_scores = get_gumbel(importance_scores) ``` is placed outside the sampling process for both training and testing, which causes the importance scores to also include Gumbel scores during testing, leading to inconsistent results every time the model is evaluated.
2. During training, the top-k tokens are output based on their scores in descending order, whereas in testing, the top-k tokens obtained via torch.topk are output in their original sequence.

**Comment by lijun2005** [comment]
I am providing a possible improvement here, where the raw mode represents the original mode.
```python



        if raw:
            importance_scores, indices = torch.topk(importance_scores, self.n_sampled_token * 2, dim=1, largest=True, sorted=True)
            importance_scores = importance_scores.transpose(1, 2).unsqueeze(-1)
            indices = indices.transpose(1, 2).unsqueeze(-1)

            q = torch.take_along_dim(q, indices=indices, dim=2)

            q_top, q_bot = torch.split(q, split_size_or_sections=[self.n_sampled_token, self.n_sampled_token], dim=2)
            imp_top, imp_bot = torch.split(
                importance_scores, 
                split_size_or_sections=[self.n_sampled_token, importance_scores.shape[2] - self.n_sampled_token],
                dim=2)

            random_indices = torch.randn(x.shape[0], self.nhead, self.n_sampled_token, 1, device=x.device).argsort(dim=2)
            q_bot = torch.take_along_dim(q_bot, indices=random_indices, dim=2)
            imp_bot = torch.take_along_dim(imp_bot, indices=random_indices, dim=2)

            q_top = SoftChooseDenseGrad.apply(imp_top, imp_bot, q_top, q_bot, self.temperature)
        else:
            importance_scores, indices = torch.topk(importance_scores, self.n_sampled_token * 2, dim=1, largest=True, sorted=True)

            # Split into two parts: first k and k-2k
            first_k_scores = importance_scores[:, :self.n_sampled_token]
            next_k_scores = importance_scores[:, self.n_sampled_token:self.n_sampled_token * 2]

            first_k_indices = indices[:, :self.n_sampled_token]
            next_k_indices = indices[:, self.n_sampled_token:self.n_sampled_token * 2]

            # Sort indices within each part to maintain the original order
            _, sorted_first_k_indices = torch.sort(first_k_indices, dim=1)
            _, sorted_next_k_indices = torch.sort(next_k_indices, dim=1)

            # Gather scores and indices back in their original order
            imp_top = first_k_scores.gather(1, sorted_first_k_indices).transpose(1, 2).unsqueeze(-1)
            first_k_indices = first_k_indices.gather(1, sorted_first_k_indices).transpose(1, 2).unsqueeze(-1)

            imp_bot = next_k_scores.gather(1, sorted_next_k_indices).transpose(1, 2).unsqueeze(-1)
            next_k_indices = next_k_indices.gather(1, sorted_next_k_indices).transpose(1, 2).unsqueeze(-1)

            q_top = torch.take_along_dim(q,indices=first_k_indices,dim=2)
            q_bot = torch.take_along_dim(q, indices=next_k_indices, dim=2)
 
            random_indices = torch.randn(x.shape[0], self.nhead, self.n_sampled_token, 1, device=x.device).argsort(dim=2)
            q_bot = torch.take_along_dim(q_bot, indices=random_indices, dim=2)
            imp_bot = torch.take_along_dim(imp_bot, indices=random_indices, dim=2)

            q_top = SoftChooseDenseGrad.apply(imp_top, imp_bot, q_top, q_bot, self.temperature)


```

**Comment by minh1409** [comment]
I'm kind of being ill so late response;

1. There is a shortcoming if not introducing noise in testing: equally important tokens break the algorithm. Consider this set: S = {a, a, a, a, a, b, b, b, b, b, c, c, c, c, c}, with a and b and c are equally important and we want to choose 3. In Eq.3, section 3.2, in the text, the criterion function takes a set and outputs a real number. In the later part of the text, we vastly simplify it by making the criterion function ~ a sum of importance scores. These random noises would increase the chance of sampling {a, b, c} out of the example.
2. The earlier variant used random indices; it breaks the locality of the chosen set of tokens; which in turn, worsens performance.
3. For added each token, we need another learnable vector paddings, which increases computational cost. From the code, if we choose q_top as top-k tokens and q_bot as the rest + a neural network distribution for paddings, it should work, maybe more performance but certainly is more expensive. O(n + k) sampling vs O(n * k) sampling.
